### ISSUE TITLE
The memory usage of the Aspnetcore program in the container is very low, and the gc is too aggressive and affects the performance

### ISSUE THREAD

**Issue by dingsongjie** [dotnet/runtime#69423]
### Description

I have an aspnetcore program running in a container with 2 cpu and 2G memory allocated，but the entire working set(MB) is only 300M in size.
![image](https://user-images.githubusercontent.com/10933723/168760953-b8932ec2-3ee4-4f62-b48f-6ee21833be14.png)
How can I make the program use more memory, reduce the frequency of gc, and finally reduce the pause time?

### Configuration

1. container: Running in a container with 2 cpu and 2G memory limited
2. aspnetcore version: Both aspnetcore 3.1 and aspnetcore 6.0
3. gc: server-gc(default)
4. System.GC.HighMemoryPercent: default

### Analysis

![image](https://user-images.githubusercontent.com/10933723/168762269-a83e8173-9f34-473e-b982-ed911e8a56d3.png)

**Comment by davidfowl** [comment]
Start here https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md?

**Comment by davidfowl** [comment]
Are you experienced full blocking GCs that affect your latency or are you observing the counters?

**Comment by Clockwork-Muse** [comment]
> make the program use more memory, reduce the frequency of gc

These goals feel counter to each other.

**Comment by mangod9** [comment]
Could you also provide more details on what kind of load was on the app? Apart from the time spent in GC are you expecting that the throughput(latency) is lower (higher) due to frequent GCs? There are various heuristics which determine GC behavior so if you could please clarify what perf you would expect and we can drill in further. Thx!

FYI, starting in 6 there is a new config around conserve memory: https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector#conserve-memory [But the default is 0, so it shouldnt affect your scenario, but mentioning in case its useful for others]

**Comment by Maoni0** [comment]
if you have a trace collected, could you please share it? the counter screenshot really doesn't say much at all. it doesn't give any kind of historical info.

**Comment by dingsongjie** [comment]
this is my nettrace file [1.nettrace](https://raw.githubusercontent.com/dingsongjie/KnowledgeBlog/main/1.rar).
This web application is built on top of ABP Framework, this application only handles some very simple requests
@Maoni0 @mangod9

**Comment by dingsongjie** [comment]
> Are you experienced full blocking GCs that affect your latency or are you observing the counters?

I see some 2N GCs in the trace file

**Comment by davidfowl** [comment]
The trace actually show about 30% of the samples spent in `String.Split` and `Encoding.GetString`:

<img width="1915" alt="image" src="https://user-images.githubusercontent.com/95136/168938286-fb8f17f0-2c77-47ce-87e3-b060ca9ff561.png">

**Comment by dingsongjie** [comment]
> Actually, is your CPU low? Can you collect a cpu sampling profile as well (with dotnet trace defaults)?

Yes the cpu usage is around 25%，is the Thread Time information on this trace file not enough? If it's not enough, I can go and collect another

**Comment by Maoni0** [comment]
it's no wonder you are doing full blocking GCs, your memory load is super high - 99% to 100%. so GC has no choice but do full blocking GCs. [this section](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#figuring-out-the-amount-of-work-for-gen2-gcs) in mem-doc explains how to find the memory load GC saw.

do you have a small memory limit set on this process?

**Comment by dingsongjie** [comment]
> it's no wonder you are doing full blocking GCs, your memory load is super high - 99% to 100%. so GC has no choice but do full blocking GCs. [this section](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#figuring-out-the-amount-of-work-for-gen2-gcs) in mem-doc explains how to find the memory load GC saw.
> 
> do you have a small memory limit set on this process?

this program running in a container with 2 cpu and 2G memory limited , there should be a lot of memory unused, and i noticed that gen0's heap is very small .https://user-images.githubusercontent.com/10933723/168787051-eddce488-259d-419c-a1e6-e7998f7e0699.png

**Comment by Maoni0** [comment]
to GC there's certainly not a lot of memory unused, because GC saw the memory load as very high, _as indicated in your trace_. 

@janvorli can you please help figure out what's consuming memory?

**Comment by davidfowl** [comment]
From looking at the profile, I see something that makes me think you are hitting a degenerate case with configuration and options. Are you or ABP using IOptionsSnapshot\<T\>?

**Comment by dingsongjie** [comment]
> From looking at the profile, I see something that makes me think you are hitting a degenerate case with configuration and options. Are you or ABP using IOptionsSnapshot<T>?

yes , i do use in some code .
![image](https://user-images.githubusercontent.com/10933723/168947785-e10c893e-582d-4b74-b7f7-44e558c2ed54.png)

**Comment by davidfowl** [comment]
This is a performance trap, it's rebinding configuration on every request leading to extremely poor performance and many allocations. Can you switch to IOptionsMonitor\<T\> instead? Then re-run your tests.

**Comment by dingsongjie** [comment]
> This is a performance trap, it's rebinding configuration on every request leading to extremely poor performance and many allocations. Can you switch to IOptionsMonitor<T> instead? Then re-run your tests.

Thanks, yes, I can switch, it does reduce memory allocation, I'll try it later.
There is another phenomenon that bothers me. According to HeapHistory, I found that the allocated memory of each heap is very small. According to this picture, the memory load has reached 98%. The gen0+gen1+gen2+GenLarge is only about 50M, and I only allocated Two cpus, due to cpu affinity, the total allocatable memory is only about 100M, which is very different from my understanding (75% * 2G). I want to know the reason for this phenomenon and how to deal with it.
![image](https://user-images.githubusercontent.com/10933723/168951298-02466938-af2d-4966-b41f-68185aa16812.png)
